Signal disk buffering (#913)
open-telemetry/opentelemetry-java-contrib: disk-buffering/README.md

SpanExporter mySpanExporter = OtlpGrpcSpanExporter.getDefault();

// Wrapping our exporter with its disk exporter.
SpanDiskExporter diskExporter = SpanDiskExporter.create(mySpanExporter, new File("/my/signals/cache/dir"), StorageConfiguration.getDefault());

 // Registering the disk exporter within our OpenTelemetry instance.
SdkTracerProvider myTraceProvider = SdkTracerProvider.builder()
        .addSpanProcessor(SimpleSpanProcessor.create(diskExporter))
        .build();
OpenTelemetrySdk.builder()
        .setTracerProvider(myTraceProvider)
        .buildAndRegisterGlobal();

```

### Reading data

Each of the exporter wrappers can read from the disk and send the retrieved data over to their
wrapped exporter by calling this method from them:

```java
try {
    if(diskExporter.exportStoredBatch(1, TimeUnit.SECONDS)) {
        // A batch was successfully exported and removed from disk. You can call this method for as long as it keeps returning true.
    } else {
        // Either there was no data in the disk or the wrapped exporter returned CompletableResultCode.ofFailure().
    }
} catch (IOException e) {
    // Something unexpected happened.
}
```

> Both, the writing and reading processes can run in parallel and the way they don't overlap is
> because each is supposed to happen in different files, and the way it is ensured that they don't
> accidentally meet in the same file is by using the configurable parameters, which ensure to set time
> frames for each action to be done on a single file at a time. On top of that, there's a mechanism in
> place to avoid overlapping on edge cases where the time frames ended but the resources haven't been
> released, for that mechanism to work properly, this tool assumes that both, the reading and the
> writing actions, are executed within the same application process.
> 

Reviewer: ```suggestion
Both the writing and reading processes can run in parallel and they don't overlap
because each is supposed to happen in different files. We ensure that reader and writer don't
accidentally meet in the same file by using the configurable parameters. These parameters set non-overlapping time frames for each action to be done on a single file at a time. On top of that, there's a mechanism in
place to avoid overlapping on edge cases where the time frames ended but the resources haven't been
released. For that mechanism to work properly, this tool assumes that both the reading and the
writing actions are executed within the same application process.

```

Reviewer: Oh and this obviously isn't a show-stopper.

> # Disk buffering

Author: It should be done now, please have a look in case the information I added about you is not accurate. Cheers.

  with the desired parameters. You can create one with default values by
  calling `StorageConfiguration.getDefault()`.

After wrapping your exporters, you must register the wrapper as the exporter you'll use. It will
take care of always storing the data it receives.

#### Set up example for spans

```java
// Creating the SpanExporter of our choice.
SpanExporter mySpanExporter = OtlpGrpcSpanExporter.getDefault();

// Wrapping our exporter with its disk exporter.
SpanDiskExporter diskExporter = new SpanDiskExporter(mySpanExporter, new File("/my/signals/cache/dir"), StorageConfiguration.getDefault());

 // Registering the disk exporter within our OpenTelemetry instance.
SdkTracerProvider myTraceProvider = SdkTracerProvider.builder()
        .addSpanProcessor(SimpleSpanProcessor.create(diskExporter))
        .build();
OpenTelemetrySdk.builder()
        .setTracerProvider(myTraceProvider)
        .buildAndRegisterGlobal();

```

### Reading data

Each of the exporter wrappers can read from the disk and send the retrieved data over to their
wrapped exporter by calling this method from them:

```java
try {

>     if(diskExporter.exportStoredBatch(1, TimeUnit.SECONDS)) {

Author: Not sure if I fully understood the question, though essentially this method is meant to be called on the consumer side whenever they want to start exporting the stored data. At the moment I didn't create any mechanism to automatically do this work, mostly because I'm thinking that, in Android, there are tools considered to be best practices for this kind of repetitive, background work such as [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager), that are preferable over going with a more generic solution of maybe spinning up a Thread with a loop inside that calls this method. So I left this uncalled on purpose so that each platform can use the mechanism that suits it best, and also to avoid making this PR bigger 😅 though I'm open to discussing an automatic read/export mechanism by default that could be optional for implementations that aren't willing or don't need to use specific optimized solutions.

> Where is the missing piece that detects that the regular (e.g. grpc) exporter fails - and only sends this part later?

Regarding this, the way it currently works is that, whenever the consumer calls the [exportStoredBatch](https://github.com/LikeTheSalad/opentelemetry-java-contrib/blob/disk-buffer/disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/exporters/AbstractDiskExporter.java#L39) method, 2 tasks are attempted inside it. The first one is to try to read a batch of signals (a JSON line) and then if it finds one, it proceeds to delegate said batch to the wrapped exporter and wait for it to finish. This is what could happen:

- If there's no data in disk, then `exportStoredBatch` will return false.
- If there's data but the wrapped exporter (for example, the grpc one) doesn't succeed when trying to export the batch, then the data is not removed from the disk and `exportStoredBatch` would return false.
- If there's data and also the wrapped exporter succeeds in exporting the batch, then the exported data is removed from the disk, and `exportStoredBatch` would return true.

So based on that information, the retry strategy would have to be decided by the consumer of this library, since they would be the ones calling this method on-demand. In the meantime, before they call this method, all it would happen is that the intercepted batches would keep on getting stored in disk, and if the configured maximum space (or [maxFolderSize](src/main/java/io/opentelemetry/contrib/disk/buffering/storage/StorageConfiguration.java#L43)) is reached, then the oldest file will be removed to make space for new data to be stored.

>     if(diskExporter.exportStoredBatch(1, TimeUnit.SECONDS)) {

Reviewer 2: I just revisited the readme and realized that I misunderstood the high-level approach.

I think I get it now - maybe you want to add paragraph - provided it's correct :smile: 

- the disk exporter is meant to write data to disk **instead** of sending it directly (i.e. it's not meant to store only failed batches to disk)
- the disk exporter doesn't include an export loop (at the moment), because that is framework specific (e.g. [android background work](https://developer.android.com/guide/background))

>     if(diskExporter.exportStoredBatch(1, TimeUnit.SECONDS)) {

Author: I agree, I'm glad someone else takes a look at the docs because is usually difficult for me to describe these things properly when I wrote the code 😅 - I'll add that information to it, cheers!

>     if(diskExporter.exportStoredBatch(1, TimeUnit.SECONDS)) {

Author: I've just added some changes to the README, I think it should be better now.

the ones provided in here:

* For a LogRecordExporter, it must be wrapped within
  a [LogRecordDiskExporter](src/main/java/io/opentelemetry/contrib/disk/buffering/exporters/LogRecordDiskExporter.java).
* For a MetricExporter, it must be wrapped within
  a [MetricDiskExporter](src/main/java/io/opentelemetry/contrib/disk/buffering/exporters/MetricDiskExporter.java).
* For a SpanExporter, it must be wrapped within
  a [SpanDiskExporter](src/main/java/io/opentelemetry/contrib/disk/buffering/exporters/SpanDiskExporter.java).

Each wrapper will need the following when instantiating them:

* The exporter to be wrapped.
* A File instance of the root directory where all the data is going to be written. The same root dir
  can be used for all the wrappers, since each will create their own folder inside it.
* An instance
  of [StorageConfiguration](src/main/java/io/opentelemetry/contrib/disk/buffering/storage/StorageConfiguration.java)
  with the desired parameters. You can create one with default values by
  calling `StorageConfiguration.getDefault()`.

After wrapping your exporters, you must register the wrapper as the exporter you'll use. It will

> take care of always storing the data it receives.

Author: You're right. I'm not fully sure but I think flushing the data might only be needed when the output stream is buffered, which was the case at first, but then I replaced it with FileOutputStream directly when I ran some tests and found out that some data was lost when the app might be terminated without closing/flushing the buffered stream before. So it seemed to me that without the buffer, each time the `write` was called, the data was written right away, which seems more reliable at least for Android.

> take care of always storing the data it receives.

Reviewer 2: great - this would be a valuable piece to add to the readme (maybe a design section)

> take care of always storing the data it receives.

Author: Thanks! I've just mentioned this in the CONTRIBUTING.md file.
